Adapt the package building scripts to use Python 3

paride opened the pull request:
Changes:
 - Do not template debian/rules as python3 is the only supported version
 - Drop six from requirements.txt
 - Makefile: drop everything related to Python 2
 - run-container: install the CI deps only on ubuntu|debian
 - read-version: update the shebang to use Python 3
 - brpm: read_dependencies(): drop unused argument
 - read-dependencies: switch to Py3 and drop the --python-version option
 - pkg-deps.json: drop the Python version field and update the redhat deps
 - Update RPM the spec file to use Python 3 when building the RPM
 - bddeb: drop support for Python 2

paride commented on tools/run-container:
```diff
@@ -463,9 +463,9 @@ main() {
-        python3 ./tools/read-dependencies "--distro=${OS_NAME}" \
-            --test-distro || {
+    local rdcmd=(python3 tools/read-dependencies "--distro=${OS_NAME}" --install)
+    [[ ${OS_NAME} =~ (ubuntu|debian) ]] && rdcmd+=('--test-distro')
```
The immediate reason why I'm passing `--test-distro` only on ubuntu|debian is that `python3-tox` is not available on CentOS.

Anyway `--test-distro` only does two thing on top on what is done without it:

 1. Adds the packages defined in `CI_SYSTEM_BASE_PKGS` to the dependencies
 2. Actually installs the packages on the system

and in my understanding is meant to enable testing of the CI automation, something we are doing only on Ubuntu anyway. The deps in `test-requirements.txt` are still installed as `brpm` and `bddeb` explicitly make `read-dependencies` read it, so the unit testing can still be done even without `--test-distro` provided that all the deps are available (not the case on CentOS, but different story).

smoser replied:
I guess if you're trying to make this work despite lack of available deps on centos (7?) then the change/knowledge to ignore should go into tools/read-dependencies.  Because if someone goes to fix this, they'll now have to remember to add the '--test-distro' flag back here, *and* fix the tools/read-dependencies to do have the right packages.

but make this is fine. thanks for explaining.

we really do need to have the ability to run test on linux other than ubuntu "really easily". Hopefully @OddBloke's pytest will land and make that possible?

paride replied:
@smoser thanks for your feedback here.

Deps are missing in CentOS 8 too (at least: `python3-tox`, `python3-contextlib2`, `python3-httpretty`, and `python3-unittest2` is only available in the PowerTools repo).

I'm not sure I agree the "install the test deps on ubuntu/debian only" logic should go in `read-dependencies`. If I leave `run-container` as it is it will always call `read-dependencies --test-distro`, which if the distro is (say) CentOS 8 can do two things:

1. Try to setup the test environment and fail, because dependencies are missing. This is consistent with everything, but not good for us, as we need to run the thing.
2. Skip setting up the test environment if distro != ubuntu|debian (or similar conditions). This is lying to the user, isn't it? A test environment has been asked for, but it hasn't been set up.

I think that not asking for a test environment (that is, not passing `--test-distro`) when we know it can't be set up is the correct thing to do.

raharper replied:
@paride Now that pytest change has landed #211 does that mean we can run tests on Centos8?  or are there still deps missing?

> I'm not sure I agree the "install the test deps on ubuntu/debian only" logic should go in read-dependencies. If I leave run-container as it is it will always call read-dependencies --test-distro, which if the distro is (say) CentOS 8 can do two things:

I'd be OK not passing --test-distro to Centos8 as long as we logged a warning that we skip it due to the fact that it's currently broken.  We should open a bug for centos8 test-enviroment  as an upstream bug and mention that in the warning.

paride replied:
There are still deps missing: contextlib2, httpretty, unittest2. Of these unittest2 is available in the PowerTools repository, the other two are not (according to pkgs.org). I'm +1 with opening a bug and mentioning it in a warning message.

paride replied:
@raharper I had a look at `https://www.rdoproject.org/what/repos/` but it seems that it only has python2 versions of the libraries we need, and no httpretty. I enabled both the openstack-train and openstack-train-testing repos, which should have the most up to date packages.

```
[root@centos8 ~]# yum repolist
Last metadata expiration check: 0:04:41 ago on Thu Mar 12 18:36:28 2020.
repo id                                                              repo name                                                                                            status
AppStream                                                            CentOS-8 - AppStream                                                                                 5103
BaseOS                                                               CentOS-8 - Base                                                                                      2107
*epel                                                                Extra Packages for Enterprise Linux 8 - x86_64                                                       4979
*epel-modular                                                        Extra Packages for Enterprise Linux Modular 8 - x86_64                                                  0
extras                                                               CentOS-8 - Extras                                                                                       3
openstack-train                                                      OpenStack Train Repository                                                                           2212
openstack-train-testing                                              OpenStack Train Testing                                                                              2236
rdo-qemu-ev                                                          RDO CentOS-7 - QEMU EV                                                                                 33

[root@centos8 ~]# yum --disablerepo="*" --enablerepo="openstack-train" list available | grep -E '(contextlib2|unittest2|httpretty)'
python2-contextlib2.noarch                            0.5.5-8.el7                   openstack-train
python2-unittest2.noarch                              1.1.0-15.el7                  openstack-train

[root@centos8 ~]# yum --disablerepo="*" --enablerepo="openstack-train-testing" list available | grep -E '(contextlib2|unittest2|httpretty)'
python2-contextlib2.noarch                            0.5.5-8.el7                   openstack-train-testing
python2-unittest2.noarch                              1.1.0-15.el7                  openstack-train-testing
```

paride commented on Makefile:
```diff
@@ -9,55 +7,34 @@ YAML_FILES+=$(shell find doc/examples -name "cloud-config*.txt" -type f )
-  echo read-version-failed)
-CODE_VERSION=$(shell $(PYVER) -c "from cloudinit import version; print(version.version_string())")
+READ_VERSION=$(shell python3 $(CWD)/tools/read-version || echo read-version-failed)
+CODE_VERSION=$(shell python3 -c "from cloudinit import version; print(version.version_string())")
```
Well, it's not great if the variable is not a constant, otherwise we just end up with a different repeated string (the variable name) and less readable code.

I'm not opposed to have `PYTHON=python3` and then using `$(PYTHON)`, but the change alone won't allow us to really switch interpreter (it's just a Makefile thing), and looks like adding a feature we don't plan to use for the moment.

smoser replied:
a well named variable is better than a repeated string in the same way that functions are better than copy and pasted code. it isnt' a feature we don't plan to use, its a feature of a programming language that we *are*  using.

paride replied:
I still don't fully agree; I will make an extreme example: in a script that calls `/bin/ls` several times, and without the immediate need to test different implementations of `ls`, would you set `LS=ls` and call `$LS` or just call `ls` directly?

Anyway I'm completely fine with having `PYTHON=python3` in this case and I'm adding it right now.

OddBloke commented on requirements.txt:
```diff
@@ -32,6 +32,3 @@ jsonpatch
 jsonschema
-
-# For Python 2/3 compatibility
-six
```
We still import six in a couple of places, I don't think we can drop this yet.

powersj commented:
@paride please address the conflicts

paride commented:
Conflicts resolved

powersj commented:
Updating the branch and removing the stale tag.

@blackboxsw this is on you to review

paride commented on Makefile:
```diff
@@ -1,38 +1,22 @@
 CWD=$(shell pwd)
-PYVER ?= $(shell for p in python3 python2; do \
-	out=$$(command -v $$p 2>&1) && echo $$p && exit; done; exit 1)
```
No need for PYVER, we support only py3

paride commented on Makefile:
```diff
@@ -1,38 +1,22 @@
-endif
-endif
+PYTHON = python3
+PIP_INSTALL := pip3 install
```
dropped all the logic around PYVER

paride commented on Makefile:
```diff
@@ -1,38 +1,22 @@
-CODE_VERSION=$(shell $(PYVER) -c "from cloudinit import version; print(version.version_string())")
+READ_VERSION=$(shell $(PYTHON) $(CWD)/tools/read-version || echo read-version-failed)
+CODE_VERSION=$(shell $(PYTHON) -c "from cloudinit import version; print(version.version_string())")
 
```
dropped all the logic around PYVER

paride commented on Makefile:
```diff
@@ -1,38 +1,22 @@
 all: check
 
-check: check_version test $(yaml)
+check: check_version test yaml
```
we did skip `yaml` on py2 (that's why it's in a variable), but as we support py3 there is no need for a variable

paride commented on Makefile:
```diff
@@ -42,20 +26,14 @@ pep8:
 
-pyflakes3:
-	@$(CWD)/tools/run-pyflakes3
-
```
dropped as the `pyflakes` target will run py3 pyflakes

paride commented on Makefile:
```diff
@@ -42,20 +26,14 @@ pep8:
-	python -m pytest -v tests/unittests cloudinit
-
-unittest3: clean_pyc
 	python3 -m pytest -v tests/unittests cloudinit
```
dropped as the `unittest` target will run the py3 unit tests

paride commented on packages/bddeb:
```diff
@@ -59,15 +59,9 @@ def run_helper(helper, args=None, strip=True):
-    else:
-        pyver = "3"
-        python = "python3"
 
```
in this file all the py2/py3 logic has been dropped

paride commented on packages/brpm:
```diff
@@ -42,7 +42,7 @@ def run_helper(helper, args=None, strip=True):
 
 
-def read_dependencies(distro, requirements_file='requirements.txt'):
+def read_dependencies(distro):
```
dropped unused argument

paride commented on packages/debian/control.in:
```diff
@@ -10,11 +10,10 @@ Standards-Version: 3.9.6
          iproute2,
          isc-dhcp-client
 Recommends: eatmydata, sudo, software-properties-common, gdisk
-XB-Python-Version: ${python:Versions}
```
I think lintian was saying `XB-Python-Version` was not need or deprecated or something, I don't remember precisely...

paride commented on packages/pkg-deps.json:
```diff
@@ -6,51 +6,21 @@
-            "3" : "python3-serial"
-         }
+         "pyyaml" : "python3-yaml",
+         "pyserial" : "python3-serial"
```
No need to distinguish between Py2 and Py3 anymore.

paride commented on packages/pkg-deps.json:
```diff
@@ -6,51 +6,21 @@
-         }
+         "httpretty" : "",
+         "unittest2" : "",
+         "contextlib2" : ""
```
These renames to the empty string are a trick to skip installing those dependencies on in systems in the "redhat" family, because the packages are missing. This was discussed in a post-standup with @raharper IIRC.

blackboxsw replied:
* I think we still need to keep these dependencies for Fedora as I can install those packages there I believe so I think we need a separate centos distro represented as it is slightly different than fedora deps.

* The empty string you set here currently doesn't actually skip installing that specific dependency we would need a comparable change to read-dependencies: translate_pip_to_system_pkg to specifically ignore our empty string rename value otherwise it actually appends a standardized package name python3-<pip_name> if rename == "".

* Also, since the time of your original writing this PR. python3-httpretty looks like it has made it into CentOS Epel repo now, so we don't have to avoid trying to install it.

I've attached a diff to my full review that contains all of my suggestions for your thoughts.

paride commented on tools/read-dependencies:
```diff
@@ -99,7 +99,7 @@ def get_parser():
         '-s', '--system-pkg-names', action='store_true', default=False,
         dest='system_pkg_names',
-        help='The name of the distro to generate package deps for.')
+        help='Generate distribution package names (python3-pkgname).')
```
The help string was wrong.

paride commented on tools/run-container:
```diff
@@ -463,9 +463,13 @@ main() {
+    else
+        rdcmd+=('--test-distro')
+    fi
+    inside_as_cd "$name" root "$cdir" "${rdcmd[@]}" || {
```
This was object of some discussion. Basically the problem is that some test dependencies are missing from CentOS, and I couldn't find a 3rd party repo with all the needed deps, so we skip them

Even with Py2 we were not running the unittests when building the RPM (fixme?), so this is not a "downgrade", but still not optimal.

blackboxsw replied:
I think dropping Py2 is good, but here I think we should install every package we know should exist, and only specifically ignore the two that are missing on CentOS.

I agree with the warning message, but we really only need to prevent running unit tests, we should try to install every dependency we can to aid in CentOS dev environment setup as best we can.  Instead of bailing on all test deps, we can specifically prevent running unittests and install all deps that we know should be available.


```suggestion
 local rdcmd=(python3 tools/read-dependencies "--distro=${OS_NAME}" --install --test-distro)
if [[ ${OS_NAME} == centos ]]; then
        error "WARNING: Can't run CentOS unit tests missing pkg deps, LP: #1867151)."
        unittest=""
fi
```

paride commented on tools/run-pyflakes3:
```diff
@@ -1,2 +0,0 @@
-#!/bin/sh
-PYTHON_VERSION=3 exec "${0%/*}/run-pyflakes" "$@"
```
Dropped in favor of run-pyflakes, that now runs py3.

paride commented:
Thanks for the review @blackboxsw, I agree on all the line with your suggestions. I applied your patch (split in 2 commits), rebased the branch resolving conflicts, and tested the actual copr build: it works! See:

https://copr.fedorainfracloud.org/coprs/g/cloud-init/cloud-init-dev/build/1356341/

blackboxsw commented on packages/pkg-deps.json:
```diff
@@ -6,48 +6,20 @@
-         "requests" : {
-            "3" : "python36-requests"
-         }
+         "unittest2" : "OMIT: LP: #1867151 Missing CentOS pkg deps",
```
I think we can drop unittest2 and contextlib2 from renames since they are no longer required in tip of master

paride replied:
+1, and the change will make differentiating between centos and redhat not really useful anymore. Should we get rid if it too?

blackboxsw commented on tools/read-dependencies:
```diff
@@ -155,27 +152,22 @@ def parse_pip_requirements(requirements_path):
         # Find a rename if present for the distro package and python version
-        rename = renames.get(pip_name, {}).get(python_ver, None)
+        rename = renames.get(pip_name, "")
+        if rename.startswith("OMIT:"):
```
I think we can drop this OMIT: functionality now as there should be no need now that unittest2 and contextlib2 are dropped from dependencies.

blackboxsw commented on tools/run-container:
```diff
@@ -463,9 +463,12 @@ main() {
-        python3 ./tools/read-dependencies "--distro=${OS_NAME}" \
-            --test-distro || {
+    local rdcmd=(python3 tools/read-dependencies "--distro=${OS_NAME}" --install --test-distro)
+    if [[ ${OS_NAME} == centos ]]; then
+        error "WARNING: Unable to install all CentOS for test dependencies (missing deps, LP: #1867151)."
+        unittest=""
+    fi
```
We should be able to drop this and allow running unittests for centos because tip has now dropped our dependency on both unittest2 and contextlib2.

I was able to pass 
/tools/run-container --source-package --unittest --artifacts=./srpm/ centos/8
...
=================== 2306 passed, 10 skipped in 54.97 seconds ===================

with this diff 
```diff
diff --git a/packages/pkg-deps.json b/packages/pkg-deps.json
index 114e921e4..800283966 100644
--- a/packages/pkg-deps.json
+++ b/packages/pkg-deps.json
@@ -17,10 +17,6 @@
       "build-requires" : [
          "python3-devel"
       ],
-      "renames" : {
-         "unittest2" : "OMIT: LP: #1867151 Missing CentOS pkg deps",
-         "contextlib2" : "OMIT: LP: #1867151 Missing CentOS pkg deps"
-      },
       "requires" : [
          "e2fsprogs",
          "iproute",
diff --git a/tools/run-container b/tools/run-container
index 31a39b069..588f09234 100755
--- a/tools/run-container
+++ b/tools/run-container
@@ -466,7 +466,6 @@ main() {
     local rdcmd=(python3 tools/read-dependencies "--distro=${OS_NAME}" --install --test-distro)
     if [[ ${OS_NAME} == centos ]]; then
         error "WARNING: Unable to install all CentOS for test dependencies (missing deps, LP: #1867151)."
-        unittest=""
     fi
     inside_as_cd "$name" root "$cdir" "${rdcmd[@]}" || {
         errorrc "FAIL: failed to install dependencies with read-dependencies"
```

paride commented:
@blackboxsw thanks for reviewing again. I dropped `unittext2` and `contextlib2`, dropped the OMIT logic, re-enabled the unittests in run-container, and rebased.
